[PowerPC] Tune AIX shared library TLS model at function level

orcguru opened the pull request:
Under some circumstance (library loaded with the main program), TLS initial-exec and local-dynamic access model can be interchanged. We could use some simple heuristic to decide use which TLS model when the variable is accessed in which function:
* If there is single TLS variable access in the function, use TLS initial-exec model.
* If there are multiple TLS variable accesses in the function, use TLS local-dynamic model.

Commit:
[PowerPC] Rename symbols references by tls-local-dynamic model

On AIX, rename those tls-local-dynamic referenced TOC symbols, so that in
following patches we can switch between different TLS models (local-dynamic
or initial-exec) for the same TLS GV in different functions within the same
module.

Commit:
[NFC] Update comment.

orcguru commented:
I will push a change to limit the rename to 64bit only.

bzEq commented:
> so that in following patches we can switch between different TLS models (local-dynamic or initial-exec) for the same TLS GV in different functions within the same module.

Could you please outline what following patches will do?

orcguru commented:
> Could you please outline what following patches will do?

Under some circumstance (library not loaded by `dlopen`), TLS initial-exec and local-dynamic access model can be interchanged. Some particular TLS variable could be accessed by different functions in the same object file. We could use some simple heuristic to decide use which TLS model when the variable is accessed in which function.

However, to make it able to mix different TLS models towards the same TLS variable in the same object file, we have to create renames. Since it seems to me renaming is fine when there is no mix, this patch is doing the rename for all TLS local-dynamic model accesses.

Commit:
WIP: need check assert

Commit:
Reset to try another approach

Commit:
Initial commit for aix-shared-library-tls-model-heuristic

bzEq commented on llvm/lib/Target/PowerPC/PPC.td:
```diff
@@ -329,6 +329,12 @@ def FeatureAIXLocalExecTLS :
                    "for 64-bit AIX">;
 
+def FeatureAIXSharedLibraryTLSModelHeuristic :
+  SubtargetFeature<"aix-shared-library-tls-model-heuristic",
```
This should be an optimization, maybe rename to `aix-shared-library-tls-model-opt`, we perform optimization based on some heuristics.

orcguru replied:
I changed the name to `aix-shared-lib-tls-model-opt`. Hope that is fine.

bzEq commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3369,6 +3369,48 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
+      DAG.getMachineFunction().getInfo<PPCFunctionInfo>();
+  if (Subtarget.hasAIXShLibTLSModelHeuristic() &&
+      !FuncInfo->isAIXFuncUseInitDone()) {
+    std::set<const GlobalValue *> TLSGV;
```
```suggestion
    SmallPtrSet<const GlobalValue *, 8> TLSGV;
```

bzEq commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3369,6 +3369,48 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
+  if (Subtarget.hasAIXShLibTLSModelHeuristic() &&
+      !FuncInfo->isAIXFuncUseInitDone()) {
+    std::set<const GlobalValue *> TLSGV;
+    for (SDNode &Node : DAG.allnodes()) {
```
One DAG is mapped to a single basic block, I notice your description is about whole function. So the size of `TLSGV` is smaller than what you really want to count.

orcguru replied:
Emm, I will update test case to expose this issue. I will propose scan over all instructions in my next update. Thank you!

amy-kwan commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3369,6 +3369,48 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
+  if (FuncInfo->isAIXFuncUseTLSLD()) {
+    LLVM_DEBUG(
+        dbgs() << DAG.getMachineFunction().getName()
+               << " function use TLS-LD model for TLS IE/LD accesses.\n");
+    Model = TLSModel::LocalDynamic;
+  } else if (FuncInfo->isAIXFuncUseTLSIE()) {
+    LLVM_DEBUG(
+        dbgs() << DAG.getMachineFunction().getName()
+               << " function use TLS-IE model for TLS IE/LD accesses.\n");
```
Minor nit on comments.
```suggestion
               << " function is using the TLS-LD model for TLS IE/LD accesses.\n");
    Model = TLSModel::LocalDynamic;
  } else if (FuncInfo->isAIXFuncUseTLSIE()) {
    LLVM_DEBUG(
        dbgs() << DAG.getMachineFunction().getName()
               << " function is using the TLS-IE model for TLS IE/LD accesses.\n");
```

amy-kwan commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3369,6 +3369,48 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
+        }
+      }
+    }
+    LLVM_DEBUG(dbgs() << format("TLSGV count:%d\n", TLSGV.size()));
```
`TLSGV.size()` is used more than once and can be pulled out.

orcguru commented on clang/lib/Basic/Targets/PPC.h:
```diff
@@ -80,6 +80,7 @@ class LLVM_LIBRARY_VISIBILITY PPCTargetInfo : public TargetInfo {
   bool IsISA3_0 = false;
   bool IsISA3_1 = false;
   bool HasQuadwordAtomics = false;
+  bool HasAIXShLibTLSModelHeuristic = false;
```
Hi Kai, Thank you for pointing out this.

I checked the code and it seems this flag is being used by backend in various places (for example: `PPCTargetLowering::LowerGlobalTLSAddressAIX`). You referred to another example that this flag is redundant, may I know the details and I will cross check with that scenario?

bzEq replied:
The variable defined here is only visible to clang and clang doesn't use this flag further. What backend sees is the feature string and set corresponding variables in backend accordingly.

bzEq replied:
Code in `PPCTargetLowering::LowerGlobalTLSAddressAIX` like
```
Subtarget.hasAIXSmallLocalExecTLS()
```
which is generated in `PPCGenSubtargetInfo.inc`, whose value is initialized by feature string `aix-small-local-exec-tls` in LLVM IR rather than the variable defined in clang.

Commit:
Fix issue/update according to comments
(1) Address issue that only the first BB get checked inside
`LowerGlobalTLSAddressAIX`, also added test case to expose the issue.
(2) Change feature name from `aix-shared-library-tls-model-heuristic`
to `aix-shared-lib-tls-model-opt`.
(3) Update naming convention in test cases to avoid numbered
identifiers.
(4) Update comments and code simplifications.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3369,6 +3369,59 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddressAIX(SDValue Op,
   bool Is64Bit = Subtarget.isPPC64();
   bool HasAIXSmallLocalExecTLS = Subtarget.hasAIXSmallLocalExecTLS();
   TLSModel::Model Model = getTargetMachine().getTLSModel(GV);
+  // Initialize TLS model opt setting lazily:
```
This primarily sets the TLS model per function. Not sure if it is possible, but can it/does it make sense for this part to be separated into its own function?

orcguru replied:
Sure. Created static function guarded by `Subtarget.hasAIXShLibTLSModelOpt()`. Thank you for the suggestion!

orcguru replied:
Sure. Created static function guarded by `Subtarget.hasAIXShLibTLSModelOpt()`. Thank you for the suggestion!

Commit:
Pull out logic as static function guarded by
`Subtarget.hasAIXShLibTLSModelOpt()`.

amy-kwan reviewed (approved):
Thank you for addressing my comment!
I think I do not have any further comments, so unless if anyone has any other comments LGTM.

Commit:
[NFC] address comment

orcguru commented:
Since https://github.com/llvm/llvm-project/pull/86641 has been merged, I will continue work on this patch to handle the interactions of these two flags.

Mean while I'm waiting for more review comments from experts. Thank you!

hubert-reinterpretcast commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3362,6 +3362,64 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddress(SDValue Op,
+                if (const GlobalValue *GV =
+                        dyn_cast<GlobalValue>(II->getOperand(0))) {
+                  TLSModel::Model GVModel = TM.getTLSModel(GV);
+                  if (GVModel == TLSModel::InitialExec ||
```
We should only count the number of `local-dynamic` accesses.

Rationale:
The option can be taken as providing the information that the shared object is loaded with the executable (not `dlopen`ed later). In that case, all variables accessed using the `local-dynamic` model can be accessed using `initial-exec`; however, there would still be variables accessed using `initial-exec` that cannot be accessed using `local-dynamic`.

hubert-reinterpretcast commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3362,6 +3362,64 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddress(SDValue Op,
+
+    unsigned TLSGVCnt = TLSGV.size();
+    LLVM_DEBUG(dbgs() << format("TLSGV count:%d\n", TLSGVCnt));
+    if (TLSGVCnt == 1)
+      FuncInfo->setAIXFuncUseTLSIE();
+    else if (TLSGVCnt > 1)
+      FuncInfo->setAIXFuncUseTLSLD();
```
We should need to set only one thing (to represent "use IE instead of LD"). The threshold number should not be repeated in more than one place.

Additionally, the threshold number may need to be increased (tests via experiments).

orcguru replied:
I added target hidden flag `PPCAIXTLSModelOptUseIEForLDLimit` in PPCISelLowering.cpp, and its initial value will be 1. Also updated test case to show the effect of change this parameter.

hubert-reinterpretcast reviewed (commented):
See inline comments for rationale re: scoping to LD -> IE only (and not performing IE -> LD).

Commit:
Update according to comments:
(1) Update Options.td to make it clear this is about the transform from LD to IE.
(2) Limit transfrom from LD to IE only.
(3) Add target hidden flag to control Limit count of the transformation.
(4) Update test case.

orcguru commented on clang/include/clang/Driver/Options.td:
```diff
@@ -5019,6 +5019,10 @@ def maix_small_local_exec_tls : Flag<["-"], "maix-small-local-exec-tls">,
+def maix_shared_lib_tls_model_opt : Flag<["-"], "maix-shared-lib-tls-model-opt">,
+  Group<m_ppc_Features_Group>,
+  HelpText<"For shared library loaded with the main program, use heuristics to "
+           "tune the TLS model at the function level (AIX 64-bit only).">;
```
I did update the description. Please let me know if another statement is better. Thank you!

Commit:
(1) Add test case to check aix-shared-lib-tls-model-opt and
aix-small-local-dynamic-tls interactions.
(2) Since TM.getTLSModel() result may need further check once this
is enabled, update GetSymbolRef() logic.
(3) Removed "heuristic" from comments.

amy-kwan commented on llvm/lib/Target/PowerPC/PPCISelLowering.cpp:
```diff
@@ -3362,6 +3367,54 @@ SDValue PPCTargetLowering::LowerGlobalTLSAddress(SDValue Op,
+      DAG.getMachineFunction().getInfo<PPCFunctionInfo>();
+  if (!FuncInfo->isAIXFuncTLSModelOptInitDone()) {
+    SmallPtrSet<const GlobalValue *, 8> TLSGV;
+    // Iterate over all instructions within current function, collect all TLS
```
nit on comment:
```suggestion
    // Iterate over all instructions within the current function and collect all TLS
```

amy-kwan reviewed (approved):
LGTM after all the updates. Thanks!
